### DISCUSSION TITLE
Example with NestJS

### DISCUSSION THREAD

**Discussion by besscroft** [nuxt/nuxt#16090] [Questions]
Hello the community,

I am looking for an example (starter) to using NestJS with Nuxt3.
Someone can point me to the good direction?

Thanks!

**Comment by wobsoriano** [comment]
Do you need 2 servers running or do you want to integrate Nuxt to Nest or vice versa?

**Reply by besscroft** [reply]
I want to put them in the same project, but want to run on the same port so that they can be accessed with one domain name.

**Comment by evanlong-me** [comment]
I want to use `Nestjs` in the `server` directory in `Nuxt3` 😉.

It is best to use it in the same way as the built-in server.

**Reply by Baroshem** [reply]
I don't recommend that.

Both frameworks should exist as separate applications (trust me, I did that experiment with building both apps frontend and backend in one repository as a part of server and it was not a good idea 😆 )

**Comment by wobsoriano** [comment]
I haven't tried Nestjs but looking at their docs, you can probably use the express instance inside the Nest application:

First, install required deps:

```bash
npm i @nestjs/common @nestjs/core @nestjs/platform-express reflect-metadata express-route-parser
```

Then update your `tsconfig`:

```json
{
  "extends": "./.nuxt/tsconfig.json",
  "compilerOptions": {
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true
  }
}
```

Create a Nitro plugin

```ts
// nitro/plugins/nest.ts

import { NestFactory } from '@nestjs/core';
import { AppModule } from '~/nest-app/app.module';
import { createRouter } from 'radix3'
import { parseExpressApp } from 'express-route-parser'

export default defineNitroPlugin(async (nitroApp) => {
  const app = await NestFactory.create(AppModule);

  await app.init()

  const router = createRouter()
  const express = app.getHttpAdapter().getInstance()

  parseExpressApp(express).forEach((route) => {
    router.insert(route.path, route)
  })
  
  nitroApp.h3App.stack.unshift({
    route: '/',
    match(url) {
      const r = router.lookup(url)
      return r !== null
    },
    handler: fromNodeMiddleware(express)
  })
})
```

```ts
export default defineNuxtConfig({
  nitro: {
    plugins: ['~/nitro/plugins/nest']
  }
})
```

**Reply by besscroft** [reply]
Thanks for the idea, I'll try it!

**Reply by wobsoriano** [reply]
While this may or may not work, as @Baroshem said in [this](https://github.com/nuxt/nuxt/discussions/16090#discussioncomment-4843478) comment, it is not recommended to mix the two.

**Comment by milenkovic** [comment]
I create a minimal starter for `nuxt3-nestjs` based on @wobsoriano proposal.

https://github.com/milenkovic/nuxt3-nestjs-starter

**Comment by Baroshem** [comment]
Having Nest.js as a part of Nuxt middleware looks actually incredible but TBH I don't recommend that.

Nuxt.js is meant to be some sort of Fullstack JS Framework, while Nest.js is supercharged Node.js/Express.js backend framework. In my work, I am building an app that has a Vue 3 (soon probably Nuxt 3) frontend and Nest.js backend with microservices architecture. And I can really recommend this stack as both frameworks are used in their respectful environments. Having Nest.js as a part of Nuxt server middlewares (nitro event handlers) might be a little bit of an overkill for handling simple requests, while if you want to dig deeper into more advanced architecture, this approach will most probably create more problems than solves 😃 

It is not a rant by any means, both frameworks are great (I am using them on the daily basis). I am just recommending to use them in the right environment to make them do what they are best at ;)

**Reply by milenkovic** [reply]
What is your proposal then for apps that have complex both frontend and backend part of the app?
So far we have used Nuxt2 + NestJS, and they worked quite nice together.

**Reply by Baroshem** [reply]
The stack you mentioned is completely fine (I am using similar in my project). 

My recommendation is just to not mix them together (like running Nest.js inside Nuxt middleware and the opposite). I am sure that if you will have two separate applications (Nuxt for frontend and Nest.js for backend) that will communicate with each other through REST or GraphQL, they will serve their purpose well :)

**Reply by milenkovic** [reply]
We actually have it in a same app(to be precise we have 3 projects running this stack), starting Nuxt as a service of Nestjs. It's was a bit of a pickle to set it up initially, but after it we had no problems.
I actually prefer having them as a 2 separate apps, but the problem was that they should run on same instance under same domain name and ideally under same port.
